### PR TITLE
Improve numeric stability for LayerNorm on CPU

### PR DESCRIPTION
**Pull request by xiaomengy**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40308 Improve numeric stability for LayerNorm on CUDA
* **#40307 Improve numeric stability for LayerNorm on CPU**

Improve numeric stability for LayerNorm on CPU.

This diff will slightly decrease the performance of LayerNorm but increase the numerical stability especially when the variance is small.

Example.

M = 1024, N = 32768
input =  torch.randn(M, N, dtype=torch.float) * 0.01 + 100.0

Previously in this case, the computed variance will be all 0s.
After this diff, the variance will be

[1.002566e-04, 9.923905e-05, 1.014871e-04, ..., 1.015831e-04, 1.000394e-04, 1.009022e-04]

This is is close to 1e-4 as we expected. The max relative difference compare to double result is 0.0005683.

The average running time for F.layer_norm with input above on one devgpu before this PR is about 99.68ms, and after this PR is about 101.32ms. So the influence on performance should be limited.

Differential Revision: [D21993646](https://our.internmc.facebook.com/intern/diff/D21993646/)

### CODE REVIEW

**Comment by BIT-silence** [comment]
link https://github.com/pytorch/pytorch/issues/40302

File: aten/src/ATen/native/cuda/layer_norm_kernel.cu [hunk 1]
```diff
@@ -328,8 +335,26 @@ void LayerNormBackwardKernelImplInternal(
       gamma.defined() ? gamma.template data_ptr<T>() : nullptr;
   T* dX_data = dX->defined() ? dX->template data_ptr<T>() : nullptr;
   cudaStream_t cuda_stream = at::cuda::getCurrentCUDAStream();
+
+  if (M == 0) {
+    if (dgamma->defined()) {
+      T* dgamma_data = dgamma->data_ptr<T>();
```
[the remark this answers is not in the document]
Comment (BIT-silence) [reply aten/src/ATen/native/cuda/layer_norm_kernel.cu:341]: Changed back to at::zeros outside.

File: aten/src/ATen/native/layer_norm.cpp [hunk 2]
```diff
@@ -27,13 +27,12 @@ std::tuple<Tensor, Tensor, Tensor> layer_norm_cpu(
   Tensor Y = at::native::empty_like(X, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
   Tensor mean = at::empty({M}, X.options());
   Tensor rstd = at::empty({M}, X.options());
-  if (M > 0) {
-    LayerNormKernel(kCPU, X, gamma, beta, M, N, eps, &Y, &mean, &rstd);
-  }
+  LayerNormKernel(
```
[the remark this answers is not in the document]
Comment (BIT-silence) [reply aten/src/ATen/native/layer_norm.cpp:30]: Reverted.

File: aten/src/ATen/native/cuda/layer_norm_kernel.cu [hunk 3]
```diff
@@ -268,6 +268,9 @@ void LayerNormKernelImplInternal(
   DCHECK_EQ(X.numel(), M * N);
   DCHECK(!gamma.defined() || gamma.numel() == N);
   DCHECK(!beta.defined() || beta.numel() == N);
```
[the remark this answers is not in the document]
Comment (BIT-silence) [reply aten/src/ATen/native/cuda/layer_norm_kernel.cu:270]: Actually they were changed in the next PR.

**Comment by xiaomengy** [comment]
https://github.com/pytorch/pytorch/pull/59987 should fix the issue

**Pull request closed without merging**